postgresql/gen-yaml/gen.sh

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+set -o errexit -o pipefail -o nounset
+
+mkdir -p manifests/upstream
+rm -f manifests/upstream/*
+
+# Fetch from helm repo: https://github.com/bitnami/charts/tree/main/bitnami/postgresql
+helm repo add bitnami https://charts.bitnami.com/bitnami
+
+helm template "postgresql" bitnami/postgresql --version "${BITNAMI_POSTGRES_RELEASE}" \
+  --set fullnameOverride="postgresql" \
+  --set nameOverride="postgresql" \
+  --set commonAnnotations."app\.uw\.systems\/description"="" \
+  --set commonAnnotations."app\.uw\.systems\/tier"="" \
+  --set commonAnnotations."app\.uw\.systems\/repos"="https://github.com/utilitywarehouse/shared-kustomize-bases/tree/main/postgresql" \
+  --set auth.existingSecret="postgresql" \
+  --set auth.secretKeys.adminPasswordKey="password" \
+  --set auth.secretKeys.userPasswordKey="user-password" \
+  --set global.postgresql.auth.database="database" \
+  --set global.postgresql.auth.username="testuser" \
+  --set primary.resources.requests.cpu="0" \
+  --set primary.resources.limits.cpu="1000m" \
+  --set primary.resources.requests.memory="0" \
+  --set primary.resources.limits.memory="1Gi" \
+  --set metrics.enabled="true" \
+  --set backup.enabled="false" > manifests/upstream/postgres.yaml
```

Comment:
Right now, it gives us option to generate only one manifest. Could you create [new directory per makefile target](https://github.com/utilitywarehouse/shared-kustomize-bases/blob/main/redis/gen-yaml/gen.sh#L29), similar to Redis, so that any team can generate their own variant of Postgres?

Reply:
Yeah that was by design. I think it would be easier to patch the single version with kustomize rather than each team having their own right?

Or maybe a better question. What would be difference between each version?

Reply:
> I think it would be easier to patch the single version with kustomize rather than each team having their own right?

Why? It's exactly the same effort to patch them. It is even more difficult if there is only one kustomize file, because team has to kustomize patch everything, instead of having sensible baiscs prepared by Makefile target.

For example, I would suggest to add the Postgres backup to the manifest generator. Then, any team using Postgres, would be able to decide whether they want the backup or not using one env variable. 
Without adding it to generator it creates much more boilerplate, is more error prone and with time the logic will start to diverge.. So, we would miss the  benefits of the current solution.

> What would be difference between each version?

For example, one team can use manifest with TLS, and another without- check out the [Kafka manifests](https://github.com/utilitywarehouse/kafka-manifests/blob/master/bitnami/Makefile), they have the same approach.
Imagine how difficult would it be with patching everything through Kustomize.

Another example- let's say, that one team wants to upgrade the chart version (e.g.  to upgrade db version), and another can't. If each team can generate their own manifest, then everyone can update their manifest however they prefer. 

I hope I was able to show you the benefits of generating as much as possible :slightly_smiling_face:

Reply:
If we are showing a golden path with this I think having two golden paths is a bad idea. Manifests here https://github.com/utilitywarehouse/cockroachdb-manifests/tree/master are a better example of what I think postgres should look like.

The kafka example is exactly my point. Everyone's Makefile command is pretty much the same but they still have to enter it there even though they don't want to change anything. @sbuliarca was saying Kafka is like that because there were issues with using kustomize not by design?

The boilerplate thing doesn't make sense to me either. You are either adding boilerplate here or in kustomize and one of them is much closer to what we have everywhere else in the company

Reply:
Kafka manifests Makefile targets are almost the same- indeed :slightly_smiling_face:  and here is where the magic happens.
Even though they are different just by few lines, check out the [diff between the files generated by two targets](https://gist.github.com/MarcinGinszt/775c2f7e1241614d9069c7361a2e1b63). 
I've got it doing `diff -bur energy/kafka-bitnami pubsub/kafka-shared `.

Generating manifests allows creating huge pieces of code with one parameter- instead of defining them each time, for each of instances.
> The boilerplate thing doesn't make sense to me either. You are either adding boilerplate here or in kustomize and one of them is much closer to what we have everywhere else in the company

With generating the manifests, you can add it once, and just reference each time. If you rather wish to add them by kustomize, you have to add them for each of instances. What's the benefit of that?

> Manifests here (...) are a better example of what I think postgres should look like.

Setting up cockroach demands adding a lot of boilerplate for each of teams. Process to do that isn't the best now. I personally think that the new approach is step forward. 

> If we are showing a golden path with this I think having two golden paths is a bad idea.

Why? Why not have two golden paths? For example, for postgres- one with backups, and one without.

Reply:
I think even with your suggestion you will still need the kustomize patches per deployment to overwrite the names etc because most teams will run more than a single instance of postgres. Or are you suggesting that each instance has it's own makefile rule?

Reply:
Yeah, I don't think we should have Makefile rule per instance. I'm not against kustomization per se.
For example, if only difference will be manifest name and resource limits, then you definitelly should kustomize. 
But let's say, you need the backup just for some of the instances- then on your place I would generate two manifests, one with backups and another without, and then Kustomize using those bases. 
Do that sounds reasonable?

Reply:
Yeah that would probably be the way I would do it too. Since they would be a bit more difficult to separate just with kustomize.

Reply:
I have split into a `with backup` and `postgres` but just using kustomize. The helm chart generates a useless backup so we still need our hardcoded yaml for now.